common/workunit/referencedfilelist.cpp

```diff
@@ -474,13 +488,213 @@ void ReferencedFile::resolve(const char *dstCluster, const char *srcCluster, IUs
+    ForEach(*targets)
+    {
+        IPropertyTree &target = targets->query();
+        if (target.hasProp("@queue"))
+            dfuQueueName.set(target.queryProp("@queue"));
+    }
+#endif
+}
+
+
+static void dfuCopy(const char *dfu_queue, StringBuffer &publisherWuid, IUserDescriptor *user, const char *sourceLogicalName, const char *destLogicalName, const char *destNodeGroupIn, const char *srcDali, bool supercopy, bool overwrite, bool preserveCompression, bool nosplit)
+{
+    if(isEmptyString(publisherWuid))
+        throw MakeStringException(-1, "Failed to create Publisher DFU Workunit.");
+    if(isEmptyString(sourceLogicalName))
+        throw MakeStringException(-1, "Source logical file not specified.");
+    if(isEmptyString(destLogicalName))
+        throw MakeStringException(-1, "Destination logical file not specified.");
+
+    PROGLOG("Copy from %s to %s", sourceLogicalName, destLogicalName);
+
+    StringBuffer destFolder, destTitle, defaultFolder, defaultReplicateFolder;
+    StringBuffer srcNodeGroup;
+    StringBuffer destNodeGroup(destNodeGroupIn);
+    if(destNodeGroup.isEmpty())
+    {
+        getNodeGroupFromLFN(user, sourceLogicalName, destNodeGroup);
+        DBGLOG("Destination node group not specified, using source node group %s", destNodeGroup.str());
+    }
+
+    CDfsLogicalFileName lfn;
+    parseLogicalPath(destLogicalName, destNodeGroup.str(), NULL, destFolder, destTitle, defaultFolder, defaultReplicateFolder);
+
+    StringBuffer fileMask;
+    constructFileMask(destTitle.str(), fileMask);
+
+    CDfsLogicalFileName logicalName;
+    logicalName.set(sourceLogicalName);
+    if (!isEmptyString(srcDali))
+    {
+        SocketEndpoint ep(srcDali);
+        if (ep.isNull())
+            throw MakeStringException(-1, "ReferencedFile Copy %s: cannot resolve SourceDali network IP from %s.", sourceLogicalName, srcDali);
+
+        logicalName.setForeign(ep,false);
+    }
+
+    Owned<IDistributedFile> file = queryDistributedFileDirectory().lookup(logicalName, user, false, false, false, nullptr, defaultPrivilegedUser);
+    if (!file)
+        throw MakeStringException(-1, "ReferencedFile failed to find file: %s", logicalName.get());
+
+    if (supercopy)
+    {
+        if (!file->querySuperFile())
+            supercopy = false;
+    }
+    else if (file->querySuperFile() && (file->querySuperFile()->numSubFiles() > 1) && isFileKey(file))
+        supercopy = true;
+
+    StringBuffer dfuQueueName(dfu_queue);
+    if (dfuQueueName.isEmpty())
+        getDefaultDFUName(dfuQueueName);
+
+    Owned<IDFUWorkUnitFactory> factory = getDFUWorkUnitFactory();
+    Owned<IDFUWorkUnit> wu = factory->createPublisherSubTask(publisherWuid);
+    wu->setJobName(destLogicalName);
+    wu->setQueue(dfuQueueName);
+
+    if (user)
+    {
+        StringBuffer username, password;
+        wu->setUser(user->getUserName(username).str());
+        wu->setPassword(user->getPassword(password).str());
```

Comment:
Where is the password coming from?  It shouldn't be available.

Reply:
Command line can still provide it, but I'll look at removing... and how it's used by DFU server.